### PR TITLE
SI-5365 Exhaustivity of extractors, guards, and unsealed traits.

### PR DESCRIPTION
**Pull request by sellout**
This builds on @retronym’s #4929.

It handles the unsealed trait case (via an additional `-Xlint:strict-unsealed-patmat` flag) and the extractor-only case (both raised by @durban).

It still defaults to lax checking by default. I don’t know what’s involved in testing it against the community build to see if we can do strict-by-default.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 1]
```diff
@@ -888,7 +893,7 @@ trait MatchAnalysis extends MatchApproximation {
               // if uniqueEqualTo contains more than one symbol of the same domain
               // then we can safely ignore these counter examples since we will eventually encounter
               // both counter examples separately
-              case _ if inSameDomain => None
+              case _ if inSameDomain => Some(WildcardExample)
```
Comment (sellout) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala]: There is probably a better way to handle this case. I’m not sure what should ever hit the previous version of this – no tests failed, but I’m sure I’m overlooking something.

**Comment by sellout** [comment]
I signed the CLA.

**Comment by retronym** [comment]
Some tests have changed the output:

From https://scala-ci.typesafe.com/job/scala-2.12.x-validate-test/3948/consoleFull, you can run this to update the expectations ("checkfiles" in partest parlance), and then see if this is as you expect.

```
# Failed test paths (this command will update checkfiles)
partest --update-check \
  test/files/neg/t1503.scala \
  test/files/neg/t5365.scala \
  test/files/neg/t5365b.scala
```

**Comment by sellout** [comment]
@retronym Thanks. I must have been running the wrong test command locally, since I missed the failures. Hopefully one of the failures will point me in the right direction for fixing my confusion in `toCounterExample`.

**Comment by sellout** [comment]
As @durban pointed out on typelevel/scala#129, these changes affect `catch` blocks as well (if `-Xlint:strict-unsealed-patmat` is used), but they don’t affect `PartialFunction`. IMO, a `catch` block should behave more like a `PartialFunction` than like an exhaustive pattern match. I’m working to make this change, but if anyone wants to jump in with a pointer, I wouldn’t mind.

**Comment by som-snytt** [comment]
I'm not following the convo, but I had an old PR to treat catches exactly as PartialFunctions.

https://github.com/scala/scala/pull/4400

**Comment by sellout** [comment]
@som-snytt Yeah, that sounds like exactly what I want. Should that PR be revived, or should I steal the part I need? (The parsing/hygiene stuff sounds a bit intimidating, so I doubt I’d pick up the whole PR, and @retronym suggested splitting it up anyway).

**Comment by milessabin** [comment]
FYI: this has been [merged](https://github.com/typelevel/scala/blob/typelevel-readme/notes/2.12.1.md#exhaustivity-of-extractors-guards-and-unsealed-traits-pull5617-sellout) in Typelevel Scala 2.12.1.

**Comment by sellout** [comment]
I see there is a failure on the “combined” check. Does that mean I should squash failing commits so that everything in the history builds cleanly?

**Comment by SethTisue** [comment]
> Does that mean I should squash failing commits so that everything in the history builds cleanly?

yes please. as a rule, before final merge we ask that every commit be green (and that unnecessary merge commits be avoided)

**Comment by durban** [comment]
Another strangeness (tested with Scalatl 2.12.1, `-Xlint:_` and `-Xstrict-patmat-analysis`):

```scala
scala> import scala.reflect.runtime.universe._
import scala.reflect.runtime.universe._

scala> (NoSymbol : Symbol) match {
     |   case NoSymbol => "nosym"
     |   case sym: Symbol if sym.isMethod => "meth"
     |   case x: Symbol => "fallback"
     | }
<console>:15: warning: match may not be exhaustive.
It would fail on the following input: (x: reflect.runtime.universe.Symbol forSome x not in NoSymbol)
       (NoSymbol : Symbol) match {
                 ^
error: No warnings can be incurred under -Xfatal-warnings.
```

However with a simple trait, it seems to work fine:

```scala
scala> trait Foo
defined trait Foo

scala> val x: Foo = new Foo {}
x: Foo = $anon$1@10bd5fb8

scala> (x : Foo) match {
     |   case `x` => "x"
     |   case foo: Foo if scala.util.Random.nextBoolean() => "rnd"
     |   case foo: Foo => "fallback"
     | }
res6: String = x

scala>
```

**Comment by sellout** [comment]
@durban Thanks. This is an interesting case. It only happens with _both_ flags enabled, unlike the other ones that only required `-Xlint:strict-unsealed-patmat`.

**Comment by sellout** [comment]
@durban `: Symbol` uses the `universe.SymbolTag.unapply` so it’s an extractor, which means it doesn’t help exhastivity. If I swap in `: SymbolApi` the warning goes away. But why does a type pattern ever use an extractor?

**Comment by durban** [comment]
@sellout Ok, thanks, that makes (some) sense. So the behavior (while somewhat surprising) seems to be correct. Those cases indeed aren't exhaustive, so the compiler is right.

As to why a (syntactically) type pattern uses an extractor, this comment on `ClassTag#unapply` seems to be the answer:
```scala
  /** A ClassTag[T] can serve as an extractor that matches only objects of type T.
   *
   * The compiler tries to turn unchecked type tests in pattern matches into checked ones
   * by wrapping a `(_: T)` type pattern as `ct(_: T)`, where `ct` is the `ClassTag[T]` instance.
   * Type tests necessary before calling other extractors are treated similarly.
   * `SomeExtractor(...)` is turned into `ct(SomeExtractor(...))` if `T` in `SomeExtractor.unapply(x: T)`
   * is uncheckable, but we have an instance of `ClassTag[T]`.
   */
```

So, if I understand correctly, this is to work around type erasure, and make more type patterns work (somewhat) safely. And it seems, that the reflection/macro API uses this feature extensively. Fair enough.

**Comment by sellout** [comment]
@durban Well, your comment just helped me understand it better than the docs I’ve been reading all morning 😆

However, I wonder if we can avoid using the extractor in the case where the type annotation is the same as the scrutinee’s type? Since then you don’t really need to use the unapply. The type doesn’t change – i.e., `(x: Q) match { ...; case y: Q => ??? }` could be rewritten to `(x: Q) match { …; case y => ??? }` safely, no?

**Comment by sellout** [comment]
In b4 @paulp explains why that doesn’t work.

**Comment by paulp** [comment]
@sellout 🤔
screen shot 2017-02-27 at 12 37 30 pm

**Comment by sellout** [comment]
@paulp http://www.urbandictionary.com/define.php?term=inb4

**Comment by paulp** [comment]
TIL.

**Comment by retronym** [comment]
@sellout The typechecker [rewrites](https://github.com/scala/scala/blob/44dac961782aa1193493c181e8423d7751c013ee/src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala#L332-L376) uncheckable type tests as class tag extractor calls before the pattern matcher reasons about whether a type test is redundant (due to the type of the scrutinee and also the knowledge that previous cases haven't matched). So the pattern matcher just sees an opaque extractor pattern and emits the `unapply` call.

In other words (inb4 @paulp again; TIAL), "an artifact of the implementation"

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 2]
```diff
@@ -761,6 +765,7 @@ trait MatchAnalysis extends MatchApproximation {
       def chop(path: Tree): List[Symbol] = path match {
         case Ident(_) => List(path.symbol)
         case Select(pre, name) => chop(pre) :+ path.symbol
+        case Apply(fun, args) => chop(fun) :+ path.symbol
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:770]: Could you please rebase this change into a separate commit to help understand its impact (ie, is a no-op without the new flag enabled?) and comment this line? I'm assuming we know must account for getting an extractor call in the tree `path` under the new option.

File: test/files/neg/patmatexhaust.scala [hunk 3]
```diff
@@ -56,7 +56,7 @@ class TestSealedExhaustive { // compile only
       case Ga =>
     }
 
-  def ma6()  = List(1,2) match { // give up
```
Comment (retronym) [inline test/files/neg/patmatexhaust.scala:59]: What change caused this test case to progress?

**Comment by som-snytt** [comment]
Isn't B4 a beginner classification in the Scala Center's update of the Odersky skill levels? The number indicates how many times you've retaken the Coursera.

**Comment by durban** [comment]
As a sidenote/question: pattern exhaustivity of destructuring `val`s (e.g., `val Some(x) = Option.empty[Int]`) doesn't seem to be checked (either with or without this PR). I guess this is because the `match` this code desugars to, seems to contain an `@unchecked` annotation. What's the reason for inserting this `@unchecked`?

**Comment by milessabin** [comment]
@sellout were you able to make any progress on @durban's issue?

**Comment by adriaanm** [comment]
Happy to reschedule for 2.12.4 once feel we can push this forward in time.

**Comment by SethTisue** [comment]
closing for inactivity. anyone want to try and carry this work forward...?

**Comment by milessabin** [comment]
I'll pick this one up too.

**Comment by adriaanm** [comment]
I'm going to close this one for now. @milessabin: happy to receive your resubmit when ready!

**Comment by SethTisue** [comment]
happy ending at https://github.com/scala/scala/pull/9140

**Comment by som-snytt** [comment]
Two follow-ups to the thread are -Xlint enables check on "destructuring vals" and the "catch takes a function" is almost in https://github.com/scala/scala/pull/4400 but I haven't looked at whether checks are applied.
